File: fruitIB.py
from typing import List

class Solution:
    def totalFruit(self, fruits: List[int]) -> int:
        count = {}
        left = 0
        max_len = 0

        for right, fruit in enumerate(fruits):
            count[fruit] = count.get(fruit, 0) + 1

            while len(count) > 2:
                left_fruit = fruits[left]
                count[left_fruit] -= 1
                if count[left_fruit] == 0:
                    del count[left_fruit]
                left += 1

            window_len = right - left + 1
            if window_len > max_len:
                max_len = window_len

        return max_len

File: fruitIB.c
int totalFruit(int* fruits, int fruitsSize) {
    int max_len = 0;
    int last_fruit = -1, second_last_fruit = -1;
    int last_fruit_count = 0;
    int current_window = 0;

    for (int i = 0; i < fruitsSize; ++i) {
        int fruit = fruits[i];

        if (fruit == last_fruit || fruit == second_last_fruit) {
            current_window++;
        } else {
            current_window = last_fruit_count + 1;
        }

        if (fruit == last_fruit) {
            last_fruit_count++;
        } else {
            last_fruit_count = 1;
            second_last_fruit = last_fruit;
            last_fruit = fruit;
        }

        if (current_window > max_len)
            max_len = current_window;
    }

    return max_len;
}
